File: Top 100 Codes/02-Working with Numbers/26-Finding Roots of a quadratic equation/FindingRootsofaquadraticequation.py
import math

def findRoots(a, b, c):
    if a == 0:
        print("Invalid equation! 'a' cannot be zero.")
        return

    d = b * b - 4 * a * c
    sqrt_val = math.sqrt(abs(d))

    if d > 0:
        print("Roots are real and distinct:")
        print((-b + sqrt_val) / (2 * a))
        print((-b - sqrt_val) / (2 * a))
    elif d == 0:
        print("Roots are real and equal:")
        print(-b / (2 * a))
    else:  # d < 0
        real_part = -b / (2 * a)
        imaginary_part = sqrt_val / (2 * a)
        print("Roots are complex:")
        print(f"{real_part} + {imaginary_part}j")
        print(f"{real_part} - {imaginary_part}j")

# Example usage
a, b, c = 1, 4, 4
findRoots(a, b, c)

File: Top 100 Codes/02-Working with Numbers/26-Finding Roots of a quadratic equation/FindingRootsofaquadraticequation.cpp
#include <iostream>
#include <cmath>

using namespace std;

void findRoots(double a, double b, double c) {
    double discriminant = b * b - 4 * a * c;

    if (discriminant > 0) {
        double root1 = (-b + sqrt(discriminant)) / (2 * a);
        double root2 = (-b - sqrt(discriminant)) / (2 * a);
        cout << "Roots are real and distinct: " << root1 << " , " << root2 << endl;
    } 
    else if (discriminant == 0) {
        double root = -b / (2 * a);
        cout << "Roots are real and equal: " << root << endl;
    } 
    else {
        double realPart = -b / (2 * a);
        double imaginaryPart = sqrt(-discriminant) / (2 * a);
        cout << "Roots are complex and imaginary: " 
             << realPart << " + " << imaginaryPart << "i , " 
             << realPart << " - " << imaginaryPart << "i" << endl;
    }
}

int main() {
    double a, b, c;
    
    cout << "Enter coefficients a, b, and c: ";
    cin >> a >> b >> c;

    if (a == 0) {
        cout << "Invalid input! 'a' cannot be zero in a quadratic equation." << endl;
    } 
    else {
        findRoots(a, b, c);
    }

    return 0;
}
